Add repository layout dev documentation

robertbindar opened the pull request:
This change adds an md file which gives a high
level overview of the scylladb/ repository, the
components each path contains and a basic description for each one of them.
This is mainly intended for onboarding engineers to help get a mental picture
when starting ramping up on Scylla concepts.

Refs #22908

regevran commented on docs/dev/repository_layout.md:
```diff
@@ -0,0 +1,216 @@
+
+Please note the source code is in an ever changing state with continuous effort being put in to refactor, decouple, isolate components. The source code hierarchy still looks flat with many source files located in the root directory, so some of these files might move in the future, some component dirs might get split, moved or removed altogether, we’ll do our best to keep the doc accurate.
+
+`alternator/`
```
```suggestion
[alternator](https://github.com/scylladb/scylladb/tree/master/alternator)
```

regevran replied:
I think that linking to `master` (in all directories) is a good idea.

regevran commented on docs/dev/repository_layout.md:
```diff
@@ -0,0 +1,216 @@
+
+`abseil/`
+
+Bundled abseil library
```
link to upstram?

regevran commented on docs/dev/repository_layout.md:
```diff
@@ -0,0 +1,216 @@
+
+`alternator/`
+
+This is the location of the Alternator project which provides interoperability with Amazon DynamoDB, here’s a Scylla Univerisity [course](https://university.scylladb.com/courses/scylla-alternator/) on it. 
```
```suggestion
The location of the Alternator project which provides interoperability with Amazon DynamoDB, here’s a Scylla Univerisity [course](https://university.scylladb.com/courses/scylla-alternator/) on it. 
```

regevran commented on docs/dev/repository_layout.md:
```diff
@@ -0,0 +1,216 @@
+
+`api/`
+
+This location contains source files managing Scylla REST APIs. ScyllaDB is being managed using the APIs defined here, the nodetool utility for instance uses such APIs to expose operations. The precursor of these APIs was Cassandra’s JMX.
```
```suggestion
Contains source files managing Scylla REST APIs. ScyllaDB is being managed using the APIs defined here, the nodetool utility for instance uses such APIs to expose operations. The precursor of these APIs was Cassandra’s JMX.
```

regevran commented on docs/dev/repository_layout.md:
```diff
@@ -0,0 +1,216 @@
+Everything related to authorization and authentication between the clients and Scylla nodes.
+
+`bin/`
+
```
Please elaborate on why do we keep binaries as part of the tree.
When are these binaries created, do we keep them in git (???), etc.

robertbindar replied:
we dont, they're just some wrapper scripts

regevran commented on docs/dev/repository_layout.md:
```diff
@@ -0,0 +1,216 @@
+
+`cmake/`
+
+This location contains all the build files related to cmake. At this moment there is an ongoing effort to migrate from configure.py based build to cmake based builds.
```
Not accurate, cmake fles are everywhere.
Here we have cmake files files that help _external_ dependencies to be used, and set global parameters  for cmake.

robertbindar replied:
cheers

regevran commented on docs/dev/repository_layout.md:
```diff
@@ -0,0 +1,216 @@
+
+`conf/`
+
+This directory contains configuration files used as defaults for running ScyllaDB locally.
```
```suggestion
Contains configuration files used as defaults for running ScyllaDB locally.
```

regevran reviewed (changes requested):
I got only to `c` :-(, will continue later.
1. Please address the comments (those that you think are meaningful) 
2. Please send a link to this PR in rnd-internal - to collect more comments.

nyh reviewed (changes requested):
I left quite a few comments (sorry...), some of them small and easy to do, in some other places I think a certain directory is not explained well and could use a better explanation.

If you don't want to polish the explanation of each of the subdirectories, please consider also the option of NOT explaining a certain directory which you can't write a good explanation for. I think this is better than including a wrong or misleading explanation for something.

By the way, how do you expect anyone to find docs/dev/repository_layout.md?
I still think putting a README.md in each subdirectory is better than a file with all the explanations, but this file will be easy to split later if we ever decide to do that.

robertbindar commented on docs/dev/repository_layout.md:
```diff
@@ -0,0 +1,217 @@
+
+[data_dictionary/](https://github.com/scylladb/scylladb/tree/master/data_dictionary)
+
+In Scylla, nodes are either `coordinator` or `replica`. data_dictionary is an abstraction on top of them so that you don’t have to use monster classes like replica database directly.
```
> The first statement is misleading in that it might suggest that we have two types of nodes, and the later statements fix this wrong impression. I think it's better to write the correct thing from the start.
> 
Agree.
> But more importantly, I didn't exactly understand why you need to mention either replica or coordinator to explain data dictionary. Wouldn't it be clearer to just explain that "data dictionary" are light-weight objects referring to the schema (keyspaces, tables, etc.) without access to the data itself (the access to data is what a "coordinator" and "replica" indeed do differently).
That's the informal explanation I initially got from Botond, I could relate to it based on what I was reading whilst onboarding, so that's why I put it here. When you onboarding, neither the concepts used in Botond's explanation nor the terms used in yours make much sense, but about coordinator/replica you read in the first documents/university courses so you can at least pin the information in your WIP mental image.
  
> 
> A longer description should also include statements like data_dictionary/data_dictionary.hh says
> 
> ```
> // No references are
> // kept, so lower-level objects like keyspaces and tables must not be held
> // across continuations.
> ```
> 
> But perhaps these details belong in data_dictionary/README.md or even in data_dictionary/data_dictionary.hh but perhaps more prominently.
Agree

robertbindar commented on docs/dev/repository_layout.md:
```diff
@@ -0,0 +1,217 @@
+
+[debug/](https://github.com/scylladb/scylladb/tree/master/debug)
+
+Assistance code for tracing/debugging (potentially unused/broken?).
```
I'll make it even simpler: https://github.com/scylladb/scylladb/issues/20981?issue=scylladb%7Cscylladb%7C23586
let's see if someone complains, I removed it from this documentation file too. I'll add back the debug/ folder when my [move](https://github.com/scylladb/scylladb/pull/23585) of debug.* into debug/ gets merged

robertbindar commented on docs/dev/repository_layout.md:
```diff
@@ -0,0 +1,217 @@
+
+Development and maintenance scripts, most notably being open-coredump.sh, a convenience script for obtaining a debugging environment with frozen toolchain.
+
+[seastar/](https://github.com/scylladb/seastar/tree/5b95d1d798b4ec12b44d7b76411c10302a2c7800)
```
It makes more sense to me. If I grep and find a match in this file, I can access the link and see a README right away, with relatives paths, I'd need to copy the path, `cd` into the subfolder, open the README.

robertbindar commented:
I addressed most of Nadav's feedback above. @amnonh @asias @Deexie @kostja if you'd like to chip in with the snippets Nadav's asked you about, I'd appreciate a lot :beers:

denesb commented:
@robertbindar ping, what is the status of this PR?

robertbindar commented:
@denesb sorry for stalling this, basically I addressed most of the comments from Nadav, it'd have been nice if the rest of the crew tagged contributed those small bits, but we can merge without and I'll file an issue to collect the info afterwards. Also I need to link the md file in the onboarding document we give to new people.

denesb commented:
Ok, I'll merge this. Incomplete documentation is better than no documentation. Improvements can come in follow-ups.
